daemons/controld/controld_cib.c

```diff
@@ -53,45 +53,46 @@ handle_cib_disconnect(gpointer user_data)
 static void
 do_cib_updated(const char *event, xmlNode * msg)
 {
+    const char *client_id = crm_element_value(msg, F_CIB_CLIENTID);
+    const char *op = crm_element_value(msg, F_CIB_OPERATION);
+    const cib__operation_t *operation = NULL;
+
+    crm_debug("Received CIB diff notification: DC=%s", pcmk__btoa(AM_I_DC));
+
     if (cib__element_in_patchset(msg, XML_CIB_TAG_ALERTS)
         || cib__element_in_patchset(msg, XML_CIB_TAG_CRMCONFIG)) {
 
         controld_trigger_config();
     }
-}
 
-static void
-do_cib_replaced(const char *event, xmlNode * msg)
-{
-    const char *client_id = crm_element_value(msg, F_CIB_CLIENTID);
-    uint32_t change_section = cib_change_section_nodes
-                              |cib_change_section_status;
-    long long value = 0;
-
-    crm_debug("Updating the CIB after a replace: DC=%s", pcmk__btoa(AM_I_DC));
     if (!AM_I_DC) {
+        // We're not in control of the join sequence
         return;
     }
 
     if (pcmk__str_eq(client_id, cib_client_id, pcmk__str_none)) {
-        // We requested this replace op. No need to restart the join.
+        // Don't restart the join sequence due to an update that we requested
         return;
     }
 
-    if ((crm_element_value_ll(msg, F_CIB_CHANGE_SECTION, &value) < 0)
-        || (value < 0) || (value > UINT32_MAX)) {
-
-        crm_trace("Couldn't parse '%s' from message", F_CIB_CHANGE_SECTION);
-    } else {
-        change_section = (uint32_t) value;
+    if (cib__get_operation(op, &operation) != pcmk_rc_ok) {
+        // Invalid operation
+        return;
     }
 
-    if (pcmk_any_flags_set(change_section, cib_change_section_nodes
-                                           |cib_change_section_status)) {
+    if (pcmk_is_set(operation->flags, cib__op_attr_replaces)
+        && (cib__element_in_patchset(msg, XML_CIB_TAG_NODES)
+            || cib__element_in_patchset(msg, XML_CIB_TAG_STATUS))) {
 
-        /* start the join process again so we get everyone's LRM status */
+        /* Nodes or status section was replaced. We assume this was triggered by
```

Comment:
comment is not quite right since transactions trigger replacements. likewise for attrd